### PR TITLE
fix: better out-of source caching and source code propagation

### PR DESCRIPTION
**Pull request by tdejager**
This adds the `SourceCodeLocation` struct that holds both the location of the manifest and the potential build-source, which is generally the actual source-code itself.

E.g given this pixi.toml (imagine it being part of a workspace with a pixi.toml):

`workspace/proj/pixi.toml`
```toml
[package]
source = "./bar"
```
Then we get:
* `manifest_source` = `workspace/proj`
* `build_source` = `./bar`

Which when called using the `source_code()` method gives you: `workspace/proj/bar`, in this specific case.

We can use this structure when to pass around and decide if we need the location to the manifest, or to the source *code*. Like when caching we need to check the source code to see if we need to do rebuilds. For example, in my plotjugger PR constant rebuilds would occur: https://github.com/facontidavide/PlotJuggler/pull/1201. This PR also fixes that behavior.

## AI Disclosure
* I found and fixed the bug as well as create the `SourceCodeLocation`, I let codex then propagate that struct throughout the codebase.

### CODE REVIEW

File: crates/pixi_command_dispatcher/src/command_dispatcher/mod.rs [hunk 1]
```diff
@@ -625,19 +625,50 @@ impl CommandDispatcher {
...
+    ) -> Result<SourceCheckout, CommandDispatcherError<SourceCheckoutError>> {
+        self.checkout_pinned_spec(pinned_spec.clone(), None).await
+    }
+
+    /// Checkout a source described by a [`SourceCodeLocation`], automatically
+    /// falling back to the manifest path when no dedicated build source is set.
+    ///
+    /// This is useful for the case where you want to checkout a source that is at
+    /// different location than the manifest, most obviously in an out-of-tree build.
+    pub async fn checkout_source_location(
```
Comment (tdejager) [inline crates/pixi_command_dispatcher/src/command_dispatcher/mod.rs]: Use this method when wanting to check out the actual source.

File: crates/pixi_command_dispatcher/src/command_dispatcher/mod.rs [hunk 2]
```diff
@@ -625,19 +625,50 @@ impl CommandDispatcher {
...
+            PinnedSourceSpec::Path(path_spec) => {
+                let alternative_root_path = match alternative_root.as_ref() {
+                    Some(PinnedSourceSpec::Path(alt_path)) => Some(
+                        self.data
+                            .resolve_typed_path(alt_path.path.to_path(), None)
+                            .map_err(|e| CommandDispatcherError::Failed(e.into()))?,
+                    ),
+                    _ => None,
+                };
+
```
Comment (tdejager) [inline crates/pixi_command_dispatcher/src/command_dispatcher/mod.rs]: This basically fixes the PlotJuggler bug, because code will be checked relative to the package manifest instead of the workspace manifest.

File: crates/pixi_command_dispatcher/src/build/mod.rs [hunk 3]
```diff
@@ -22,12 +22,88 @@ pub use dependencies::{
...
+}
+
+impl From<PinnedSourceSpec> for SourceCodeLocation {
+    fn from(manifest_source: PinnedSourceSpec) -> Self {
+        Self {
+            manifest_source,
+            build_source: None,
+        }
+    }
+}
```
Comment (remimimimimi) [inline crates/pixi_command_dispatcher/src/build/mod.rs]: I think it makes sense to introduce From only when there's only one canonical conversion possible. Here we can implement conversion for manifest_source as you did or for build_source. So I think it would be better to use explicit method.

File: crates/pixi_command_dispatcher/src/build_backend_metadata/mod.rs [hunk 4]
```diff
@@ -103,7 +100,8 @@ impl BuildBackendMetadataSpec {
         skip_all,
         name="backend-metadata",
         fields(
-            source = %self.manifest_source,
+            manifest_source = %self.source.manifest_source(),
+            build_source = ?self.source.build_source(),
```
Comment (remimimimimi) [inline crates/pixi_command_dispatcher/src/build_backend_metadata/mod.rs]: Maybe use `source = %self.source` instead?

File: crates/pixi_command_dispatcher/src/source_build/mod.rs [hunk 5]
```diff
@@ -124,7 +120,8 @@ impl SourceBuildSpec {
         skip_all,
         name = "source-build",
         fields(
-            source= %self.manifest_source,
+            manifest_source = %self.source.manifest_source(),
+            build_source = ?self.source.build_source(),
```
Comment (remimimimimi) [inline crates/pixi_command_dispatcher/src/source_build/mod.rs:124]: Same thing here

File: crates/pixi_command_dispatcher/src/source_metadata/mod.rs [hunk 6]
```diff
@@ -64,7 +59,8 @@ impl SourceMetadataSpec {
         skip_all,
         name = "source-metadata",
         fields(
-            source= %self.backend_metadata.manifest_source,
+            manifest_source= %self.backend_metadata.source.manifest_source(),
+            build_source=?self.backend_metadata.source.build_source(),
```
Comment (remimimimimi) [inline crates/pixi_command_dispatcher/src/source_metadata/mod.rs]: Same here

File: crates/pixi_command_dispatcher/src/build_backend_metadata/mod.rs (hunk 4 above)
Comment (tdejager) [reply crates/pixi_command_dispatcher/src/build_backend_metadata/mod.rs]: its optional alas.

File: crates/pixi_command_dispatcher/src/command_dispatcher/mod.rs [hunk 7]
```diff
@@ -625,19 +625,50 @@ impl CommandDispatcher {
...
+    async fn checkout_pinned_spec(
         &self,
         pinned_spec: PinnedSourceSpec,
+        alternative_root: Option<PinnedSourceSpec>,
     ) -> Result<SourceCheckout, CommandDispatcherError<SourceCheckoutError>> {
         match pinned_spec {
-            PinnedSourceSpec::Path(ref path) => {
+            PinnedSourceSpec::Path(path_spec) => {
+                let alternative_root_path = match alternative_root.as_ref() {
```
Comment (baszalmstra) [inline crates/pixi_command_dispatcher/src/command_dispatcher/mod.rs]: I think this can be replaced with `SourceAnchor`! Its purpose is to make one sourcespec relative to another.

You can also have a git source with a subdirectory for instance. I dont think the current implementation deals with that?

Comment (tdejager) [reply crates/pixi_command_dispatcher/src/command_dispatcher/mod.rs]: I think it does because there is a test for it, in the build suite and that will never be interpreted as relative.

But I didn't know that had that purpose, should I make the change and also update SourceAnchor documentation?

Comment (baszalmstra) [reply crates/pixi_command_dispatcher/src/command_dispatcher/mod.rs]: Yes please! If you dont have time I can also do it!

**Comment by baszalmstra** [comment]
I changed the code significantly. I'm trying to make sure that the pinned `build_source_spec` is always completely resolved and doesn't depend on the `manifest_source` anymore. That way, when dealing with a pinned `build_source_spec`, we can just use it as is without having to convert it in different locations.

@tdejager What is the best way to test this?

**Comment by tdejager** [comment]
There was a PR on the suite I made: https://github.com/prefix-dev/pixi-build-testsuite/pull/91

And for a real-world test you can try my PR on PlotJuggler. Can find the link if you want but I'm on mobile.

Note that there are also some unit test in pixi that test parts of this, but does not call any builds.

**Comment by baszalmstra** [comment]
I changed the code to always make the `source_build` `PinnedSourceSpec` relative to the workspace instead of to some undefined root. This works well, but it looks very strange in the lockfile. If you have a git dependency which has a out-of-source build the entire git url is repeated in the `package_build_source` field. I think it makes sense to store the path relative to the `location` only in the lock-file.

**Comment by tdejager** [comment]
I looked at the changes :) is now the preferred_source an alternative to the build_source or is it only used for the pinning during a pixi update?

File: crates/pixi_record/src/source_record.rs [hunk 8]
```diff
@@ -55,62 +55,86 @@ pub struct InputHash {
...
-}
 
-impl TryFrom<CondaSourceData> for SourceRecord {
-    type Error = ParseLockFileError;
+    /// Create SourceRecord from CondaSourceData with paths resolved relative to workspace_root.
+    /// This should be used when reading from the lock file.
+    ///
+    /// TODO: Currently this doesn't use workspace_root and just resolves build_source
+    /// relative to manifest_source. Future implementation should resolve both against
+    /// workspace_root first.
```
Comment (baszalmstra) [inline crates/pixi_record/src/source_record.rs]: This still needs implementing. The idea is: `PinnedSourceSpec` is always relative to the workspace root. But the `build_source` in the lock-file is relative to the `location` of the package. We need to convert these two. My reasoning was that this can only reliably be done if all relative paths have a base so you can compare apples to apples. The `workspace_root` should provide this base.

File: crates/pixi_record/src/source_record.rs [hunk 9]
```diff
@@ -55,62 +55,86 @@ pub struct InputHash {
...
-                    GitReference::Rev(_) => Some(GitShallowSpec::Rev),
-                    GitReference::DefaultBranch => None,
-                };
+impl SourceRecord {
+    /// Convert to CondaSourceData with paths made relative to workspace_root.
+    /// This should be used when writing to the lock file.
+    ///
+    /// TODO: Currently this doesn't use workspace_root and just makes build_source
+    /// relative to manifest_source. Future implementation should make both relative
+    /// to workspace_root first.
```
Comment (baszalmstra) [inline crates/pixi_record/src/source_record.rs]: This still needs to happen. See my other comments.

File: crates/pixi_command_dispatcher/src/source_build/mod.rs [hunk 10]
```diff
@@ -236,14 +235,16 @@ impl SourceBuildSpec {
             .await
             .map_err_with(SourceBuildError::Discovery)?;
 
+        tracing::error!("{:?}", discovered_backend.init_params);
```
Comment (baszalmstra) [inline crates/pixi_command_dispatcher/src/source_build/mod.rs]:
```suggestion
```

File: crates/pixi_record/src/path_utils.rs [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+        pathdiff::diff_paths(&target_abs, &base_abs).unwrap_or_else(|| target_abs.clone());
+    let normalized = normalize_path(&relative);
+    if normalized.as_os_str().is_empty() {
+        None
+    } else {
+        Some(unixify_path(&normalized))
+    }
+}
+
+/// Apply a manifest subdir back onto a relative path stored in the lock, again emitting `/`.
```
Comment (baszalmstra) [inline crates/pixi_record/src/path_utils.rs]: Can you explain this comment some more? Its unclear to me what this does. 

Maybe also add more edge case tests?

File: crates/pixi_record/src/path_utils.rs [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+    for part in parts {
+        normalized.push(part);
+    }
+
+    normalized
+}
+
+/// Compute the repo-relative path from `base` (manifest subdir) to `target` (build subdir), always
+/// returning `/`-separated strings so the lock format is stable across platforms.
+pub(crate) fn relative_repo_subdir(base: &str, target: &str) -> Option<String> {
```
Comment (baszalmstra) [inline crates/pixi_record/src/path_utils.rs]: Does it make sense to add more tests for this to also test some more edge cases?

File: crates/pixi_record/src/pinned_source.rs [hunk 13]
```diff
@@ -129,6 +130,152 @@ impl PinnedSourceSpec {
...
+    /// or both are git sources pointing to the same repository.
+    /// This is useful for making `build_source` relative to `manifest_source` in lock files.
+    ///
+    /// Returns `None` if:
+    /// - Not a compatible combination (different types or different git repos)
+    /// - The sources cannot be made relative to each other
+    ///
+    /// # Arguments
+    /// * `base` - The base pinned source to make this path relative to (typically the manifest_source)
+    pub fn make_relative_to(&self, base: &PinnedSourceSpec, workspace_root: &Path) -> Option<Self> {
```
Comment (baszalmstra) [inline crates/pixi_record/src/pinned_source.rs]: Maybe this should return a relative path instead of `Self`. The function only ever returns `None` or a relative path.

Comment (baszalmstra) [reply crates/pixi_record/src/pinned_source.rs]: Or even a `Utf8UnixPathBuf`?

File: crates/pixi_record/src/pinned_source.rs [hunk 14]
```diff
@@ -1013,4 +1161,439 @@ mod tests {
             SourceMismatchError::GitSubdirectoryMismatch { .. }
         ));
     }
+
+    mod matches_source_spec {
+        use super::*;
+        use pixi_spec::{PathSourceSpec, SourceLocationSpec, SourceSpec, UrlSourceSpec};
+        use typed_path::Utf8TypedPathBuf;
+
+        use crate::{PinnedPathSpec, PinnedSourceSpec, PinnedUrlSpec};
```
Comment (baszalmstra) [inline crates/pixi_record/src/pinned_source.rs]: Dont really need a submodule I think.

File: crates/pixi_record/src/pinned_source.rs [hunk 15]
```diff
@@ -1013,4 +1161,439 @@ mod tests {
...
+
+            // Should match - GitHub URLs are case-insensitive
+            assert!(pinned.matches_source_spec(&spec));
+        }
+    }
+
+    mod make_relative_to {
+        use std::path::Path;
+
+        use crate::{PinnedPathSpec, PinnedSourceSpec};
```
Comment (baszalmstra) [inline crates/pixi_record/src/pinned_source.rs]: Not needing a submodule I think.

File: crates/pixi_record/src/source_record.rs [hunk 16]
```diff
@@ -55,80 +64,129 @@ pub struct InputHash {
...
+                        let mut subdirectory = git_spec.source.subdirectory.clone();
+
+                        if let PinnedSourceSpec::Git(manifest_git) = &self.manifest_source {
+                            if same_git_checkout(&git_spec, manifest_git) {
+                                subdirectory = relative_repo_subdir(
+                                    manifest_git.source.subdirectory.as_deref().unwrap_or(""),
+                                    git_spec.source.subdirectory.as_deref().unwrap_or(""),
+                                );
+                            }
+                        }
```
Comment (baszalmstra) [inline crates/pixi_record/src/source_record.rs]: This is not needed. A git reference is already "absolute", we can keep it that way.

File: crates/pixi_record/src/path_utils.rs (hunk 11 above)
Comment (tdejager) [reply crates/pixi_record/src/path_utils.rs]: Remove all of this.

File: crates/pixi_record/src/path_utils.rs (hunk 12 above)
Comment (tdejager) [reply crates/pixi_record/src/path_utils.rs]: Remove all of this.

File: crates/pixi_record/src/pinned_source.rs (hunk 13 above)
Comment (tdejager) [reply crates/pixi_record/src/pinned_source.rs]: Yeah did just that.

File: crates/pixi_record/src/pinned_source.rs [hunk 17]
```diff
@@ -129,6 +130,249 @@ impl PinnedSourceSpec {
...
+        base: &PinnedSourceSpec,
+        workspace_root: &Path,
+    ) -> Option<Utf8UnixPathBuf> {
+        match (self, base) {
+            // Path-to-Path: Make the path relative
+            (PinnedSourceSpec::Path(this_path), PinnedSourceSpec::Path(base_path)) => {
+                let this_path = this_path.resolve(workspace_root);
+                let base_path = base_path.resolve(workspace_root);
+
+                dbg!(&this_path, &base_path);
```
Comment (baszalmstra) [inline crates/pixi_record/src/pinned_source.rs]: dbg! statemetns

File: crates/pixi_record/src/path_utils.rs [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+        normalized.push(std::path::MAIN_SEPARATOR.to_string());
+    }
+    for part in parts {
+        normalized.push(part);
+    }
+
+    normalized
+}
+
+pub(crate) fn unixify_path(path: &Path) -> String {
```
Comment (baszalmstra) [inline crates/pixi_record/src/path_utils.rs]: Maybe we can return a `Utf8UnixPathBuf` to type this type.

Comment (tdejager) [reply crates/pixi_record/src/path_utils.rs]: Ok done!

**Review by baszalmstra — APPROVED** [review]

**Comment by baszalmstra** [comment]
Nooo! It look like the windows test snapshots contain backslashes!

**Comment by tdejager** [comment]
> Nooo! It look like the windows test snapshots contain backslashes!

Pffff, it was not converting correctly. Hope I pushed the fix now. This PR Is getting on my nerves :P
